### PR TITLE
Add new Append API with DataVerificationInfo to Env WritableFile

### PR DESCRIPTION
**Pull request by zhichao-cao**
Add the new Append and PositionedAppend API to env WritableFile. User is able to benefit from the write checksum handoff API when using the legacy Env classes. FileSystem already implemented the checksum handoff API.

test plan: make check, added new unit test.

### CODE REVIEW

File: include/rocksdb/env.h [hunk 1]
```diff
@@ -772,6 +772,13 @@ class RandomAccessFile {
   // RandomAccessFileWrapper too.
 };
 
+// A data structure brings the data verification information, which is
+// used togther with data being written to a file.
+struct DataVerificationInfo {
+  // checksum of the data being written.
+  Slice checksum;
+};
+
```
Comment (mrambacher) [inline include/rocksdb/env.h:781]: Can you just leave this in file_system and make it a forward declaration?

**Comment by mrambacher** [comment]
Why do we want to add this to the legacy env rather than encouraging the new interface for this functionality?  It seems backwards to add this functionality to code that should really be deprecated.

**Review by mrambacher — COMMENTED** [review]
I do not understand why we want this functionality in a legacy interface rather than encouraging development on the new interface instead.  Why do we want to support this in an old interface that should be deprecated?

**Comment by zhichao-cao** [comment]
> I do not understand why we want this functionality in a legacy interface rather than encouraging development on the new interface instead. Why do we want to support this in an old interface that should be deprecated?

Originally, we just want to support it in File System interfaces. However, it is an urgent requests from internal products who want to use the handoff function. However, they are still based on Env and will not finish the migration to File System within 1-2 years.

File: include/rocksdb/env.h (hunk 1 above)
Comment (zhichao-cao) [reply include/rocksdb/env.h:781]: Yes. I agree, it will be easier for future deprecation from Env.

File: include/rocksdb/env.h [hunk 2]
```diff
@@ -801,6 +802,18 @@ class WritableFile {
...
+  // Append data with verification information.
+  // Note that this API change is experimental and it might be changed in
+  // the future. Currently, RocksDB only generates crc32c based checksum for
+  // the file writes when the checksum handoff option is set.
+  // Expected behavior: if currently ChecksumType::kCRC32C is not supported by
+  // WritableFile, the information in DataVerificationInfo can be ignored
+  // (i.e. does not perform checksum verification).
+  virtual Status Append(const Slice& data,
+                        const DataVerificationInfo& /* verification_info */) {
+    return Append(data);
```
Comment (anand1976) [inline include/rocksdb/env.h:814]: The ```verification_info``` will be silently ignored if the ```Env``` doesn't support it. Its the same behavior as ```FSWritableFile```, so I guess its ok.

**Review by anand1976 — APPROVED** [review]
LGTM. I agree its not an ideal to have these new APIs in Env, but we have to make the exception.

**Comment by zhichao-cao** [comment]
> LGTM. I agree its not an ideal to have these new APIs in Env, but we have to make the exception.

Right. I think once all the services migrate to the new FileSystem APIs, we may deprecate it in the future.

**Pull request closed without merging**